frontend/src/app/titles.ts

```diff
@@ -0,0 +1,121 @@
+    titleFromData: (data: QueryData) => string,
+): ResolveFn<string> => {
+    return (route: ActivatedRouteSnapshot) => {
+        const query = queryFromParams(route.params);
+        return inject(Apollo).watchQuery({ query }).valueChanges.pipe(
+            map(result => result.data),
+            map(titleFromData),
+            map(pageTitle),
+        );
+    }
+}
+
+type SourceTitleQueryData = { source?: { name: string } }
+
+const sourceTitleQuery = (params: Params) => gql<SourceTitleQueryData, unknown>(`
+    query SourceTitleQuery {
+        source(id: "${params['id']}") {
+            id
+            name
+        }
+}`);
+
+const sourceFormTitle = (data: SourceTitleQueryData) => `Edit ${data.source?.name}`;
+
+export const sourceFormTitleResolver = queryTitleResolver(
+    sourceTitleQuery, sourceFormTitle);
+
+type EntityDescriptionTitleQueryData<Key extends string> =
+    Record<Key, { name: string, source: { name: string } }>;
+
+type AgentTitleQueryData = EntityDescriptionTitleQueryData<'agentDescription'>;
+
+const agentTitleQuery = (params: Params) => gql<AgentTitleQueryData, unknown>(`
+    query AgentTitleQuery {
+        agentDescription(id: "${params['id']}") {
+            id
+            name
+            source { id, name }
+        }
+}`);
+
+const entityDescriptionFormTitle = <Key extends string>(
+    data: EntityDescriptionTitleQueryData<Key>
+) => {
+    const key = _.first(_.keys(data)) as Key;
+    const entity = data[key];
+    if (entity) {
+        return `Edit ${entity.name} (${entity.source.name})`;
+    } else {
+        return 'Not found';
+    }
+}
+
+export const agentFormTitleResolver = queryTitleResolver(
+    agentTitleQuery, entityDescriptionFormTitle);
+
+type LetterTitleQueryData = EntityDescriptionTitleQueryData<'letterDescription'>;
+
+const letterTitleQuery = (params: Params) => gql<LetterTitleQueryData, unknown>(`
+    query LetterTitleQuery {
+        letterDescription(id: "${params['id']}") {
+            id
+            name
+            source { id, name }
+        }
+    }
+`);
+
+
+export const letterFormTitleResolver = queryTitleResolver(
+    letterTitleQuery, entityDescriptionFormTitle);
+
+
+type GiftTitleQueryData = EntityDescriptionTitleQueryData<'giftDescription'>;
+
+const giftTitleQuery = (params: Params) => gql<GiftTitleQueryData, unknown>(`
+    query GiftTitleQuery {
+        giftDescription(id: "${params['id']}") {
+            id
+            name
+            source { id, name }
+        }
+    }
+`);
+
+
+export const giftFormTitleResolver = queryTitleResolver(
+    giftTitleQuery, entityDescriptionFormTitle);
+
+type SpaceTitleQueryData = EntityDescriptionTitleQueryData<'spaceDescription'>;
+
+const spaceTitleQuery = (params: Params) => gql<SpaceTitleQueryData, unknown>(`
+    query SpaceTitleQuery {
+        spaceDescription(id: "${params['id']}") {
+            id
+            name
+            source { id, name }
```

Comment:
I'd prefer it if each property were on a new line. It's not required by GraphQL in any way, but it's common practice.